File: Tutorials/Codes/System_Dynamics_Course/Chapter20/Lesson2/Chapter20_Lesson2.m
% Chapter20_Lesson2.m
% System Dynamics — Chapter 20 (Chaos, Complex Dynamics, and Computational Tools)
% Lesson 2: Bifurcation Diagrams, Period Doubling, and Strange Attractors
%
% Outputs:
%   - logistic_bifurcation_matlab.png
%   - lorenz_attractor_matlab.png
%
% Notes:
%   (1) The bifurcation diagram is built by sweeping r, discarding transients,
%       and plotting the last iterates for each r.
%   (2) The Lorenz trajectory is integrated with ode45 (continuous-time ODE).
%   (3) A minimal Simulink model can be generated programmatically (optional).

clear; clc; close all;

%% Part A — Logistic map bifurcation diagram
rmin = 2.5; rmax = 4.0; Nr = 6000;
nTransient = 1200; nKeep = 200;
rs = linspace(rmin, rmax, Nr);
x = 0.123456 * ones(1, Nr);

for k = 1:nTransient
    x = rs .* x .* (1 - x);
end

Rplot = zeros(1, Nr*nKeep);
Xplot = zeros(1, Nr*nKeep);
idx = 1;
for k = 1:nKeep
    x = rs .* x .* (1 - x);
    Rplot(idx:idx+Nr-1) = rs;
    Xplot(idx:idx+Nr-1) = x;
    idx = idx + Nr;
end

figure('Position',[100 100 900 450]);
plot(Rplot, Xplot, '.', 'MarkerSize', 1);
xlabel('r'); ylabel('x (asymptotic samples)');
title('Logistic map bifurcation diagram');
grid on;
saveas(gcf, 'logistic_bifurcation_matlab.png');

%% Part B — Lorenz attractor (ODE) with ode45
sigma = 10; rho = 28; beta = 8/3;
f = @(t, y) [ sigma*(y(2)-y(1));
              y(1)*(rho - y(3)) - y(2);
              y(1)*y(2) - beta*y(3) ];

tspan = [0 40];
y0 = [1; 1; 1];
opts = odeset('RelTol',1e-8,'AbsTol',1e-10);
[t, Y] = ode45(f, tspan, y0, opts);

% discard transient
mask = t >= 5;
Y2 = Y(mask, :);

figure('Position',[120 120 700 600]);
plot3(Y2(:,1), Y2(:,2), Y2(:,3), 'LineWidth', 0.5);
xlabel('x'); ylabel('y'); zlabel('z');
title('Lorenz attractor trajectory (ode45)');
grid on; view(45, 25);
saveas(gcf, 'lorenz_attractor_matlab.png');

%% Part C — Optional: programmatically build a simple Simulink map model
% This creates a discrete-time logistic map with a Unit Delay and MATLAB Function.
% Requires Simulink installed. If not available, this section will safely skip.
try
    hasSimulink = license('test','Simulink');
catch
    hasSimulink = false;
end

if hasSimulink
    mdl = 'Chapter20_Lesson2_LogisticMap_Simulink';
    new_system(mdl); open_system(mdl);

    add_block('simulink/Sources/Constant', [mdl '/r'], 'Value', '3.7', ...
              'Position', [50 80 90 110]);
    add_block('simulink/Sources/Constant', [mdl '/x0'], 'Value', '0.2', ...
              'Position', [50 160 90 190]);

    add_block('simulink/Discrete/Unit Delay', [mdl '/UnitDelay'], ...
              'InitialCondition', '0.2', 'SampleTime', '1', ...
              'Position', [170 150 220 200]);

    add_block('simulink/User-Defined Functions/MATLAB Function', [mdl '/LogisticFcn'], ...
              'Position', [280 130 420 220]);

    % Set MATLAB Function code
    fcn = [ ...
        'function x_next = fcn(x, r)\n' ...
        '%#codegen\n' ...
        'x_next = r * x * (1 - x);\n' ...
        'end\n' ];
    set_param([mdl '/LogisticFcn'], 'Script', fcn);

    add_block('simulink/Sinks/Scope', [mdl '/Scope'], ...
              'Position', [480 150 520 190]);

    % Wire: x0 -> UnitDelay input (sets IC only; still connect for convenience)
    add_line(mdl, 'x0/1', 'UnitDelay/1');
    add_line(mdl, 'UnitDelay/1', 'LogisticFcn/1');
    add_line(mdl, 'r/1', 'LogisticFcn/2');
    add_line(mdl, 'LogisticFcn/1', 'UnitDelay/1', 'autorouting', 'on');
    add_line(mdl, 'UnitDelay/1', 'Scope/1');

    set_param(mdl, 'StopTime', '200', 'Solver', 'FixedStepDiscrete');
    save_system(mdl);
    % run with: sim(mdl)
end

disp('Done.');

File: Tutorials/Codes/System_Dynamics_Course/Chapter20/Lesson2/Chapter20_Lesson2.py
# Chapter20_Lesson2.py
# System Dynamics — Chapter 20 (Chaos, Complex Dynamics, and Computational Tools)
# Lesson 2: Bifurcation Diagrams, Period Doubling, and Strange Attractors
#
# Outputs:
#   - logistic_bifurcation.png (bifurcation diagram)
#   - logistic_bifurcation.csv (points used in the diagram)
#   - lorenz_attractor.png (3D trajectory projection)
#   - lorenz_traj.csv (Lorenz trajectory samples)

import numpy as np
import matplotlib.pyplot as plt

def logistic_map(x, r):
    return r * x * (1.0 - x)

def bifurcation_logistic(r_min=2.5, r_max=4.0, n_r=6000,
                         n_transient=1200, n_keep=200, x0=0.123456):
    rs = np.linspace(r_min, r_max, n_r)
    x = np.full(n_r, x0, dtype=float)

    # Burn-in / transient
    for _ in range(n_transient):
        x = logistic_map(x, rs)

    # Keep last n_keep iterates for each r
    r_out = np.repeat(rs, n_keep)
    x_out = np.empty(n_r * n_keep, dtype=float)
    idx = 0
    for _ in range(n_keep):
        x = logistic_map(x, rs)
        x_out[idx:idx+n_r] = x
        idx += n_r

    # Save CSV (r, x)
    data = np.column_stack([r_out, x_out])
    np.savetxt("logistic_bifurcation.csv", data, delimiter=",", header="r,x", comments="")

    # Plot
    plt.figure(figsize=(10, 5))
    plt.plot(r_out, x_out, ",", markersize=1)
    plt.xlabel("r")
    plt.ylabel("x (asymptotic samples)")
    plt.title("Logistic map bifurcation diagram")
    plt.tight_layout()
    plt.savefig("logistic_bifurcation.png", dpi=200)
    plt.close()

def rk4_step(f, t, y, h):
    k1 = f(t, y)
    k2 = f(t + 0.5*h, y + 0.5*h*k1)
    k3 = f(t + 0.5*h, y + 0.5*h*k2)
    k4 = f(t + h, y + h*k3)
    return y + (h/6.0)*(k1 + 2*k2 + 2*k3 + k4)

def lorenz_rhs(sigma=10.0, rho=28.0, beta=8.0/3.0):
    def f(t, y):
        x, yv, z = y
        return np.array([
            sigma*(yv - x),
            x*(rho - z) - yv,
            x*yv - beta*z
        ], dtype=float)
    return f

def simulate_lorenz(T=40.0, h=0.005, y0=(1.0, 1.0, 1.0),
                    sigma=10.0, rho=28.0, beta=8.0/3.0,
                    transient=5.0, sample_stride=4):
    f = lorenz_rhs(sigma, rho, beta)
    n = int(T / h)
    y = np.array(y0, dtype=float)
    t = 0.0

    # Run and store after transient
    traj = []
    for i in range(n):
        y = rk4_step(f, t, y, h)
        t += h
        if t >= transient and (i % sample_stride == 0):
            traj.append((t, y[0], y[1], y[2]))
    traj = np.array(traj, dtype=float)
    np.savetxt("lorenz_traj.csv", traj, delimiter=",", header="t,x,y,z", comments="")

    # Plot 3D trajectory (requires mpl_toolkits)
    from mpl_toolkits.mplot3d import Axes3D  # noqa: F401
    fig = plt.figure(figsize=(7, 6))
    ax = fig.add_subplot(111, projection="3d")
    ax.plot(traj[:,1], traj[:,2], traj[:,3], linewidth=0.4)
    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.set_zlabel("z")
    ax.set_title("Lorenz attractor trajectory (sampled)")
    plt.tight_layout()
    plt.savefig("lorenz_attractor.png", dpi=200)
    plt.close()

if __name__ == "__main__":
    bifurcation_logistic()
    simulate_lorenz()
    print("Done. Generated PNGs and CSVs in the current directory.")
